modules/python/common.cmake

```diff
@@ -230,6 +230,30 @@ if(NOT OPENCV_SKIP_PYTHON_LOADER)
   install(FILES "${__python_loader_install_tmp_path}/cv2/${__target_config}" DESTINATION "${OPENCV_PYTHON_INSTALL_PATH}/cv2/" COMPONENT python)
 endif()  # NOT OPENCV_SKIP_PYTHON_LOADER
 
+# Install pip metadata files to ensure that opencv installed via CMake is listed by pip list
+# See https://packaging.python.org/specifications/recording-installed-packages/
+# and https://packaging.python.org/en/latest/specifications/core-metadata/#core-metadata
+option(OPENCV_PYTHON_PIP_METADATA_INSTALL "Use CMake to install Python pip metadata. Set to off if some other tool already installs it." OFF)
+mark_as_advanced(OPENCV_PYTHON_PIP_METADATA_INSTALL)
+set(OPENCV_PYTHON_PIP_METADATA_INSTALLER "cmake" CACHE STRING "Specify the string to identify the pip Installer. Default: cmake, change this if you are using another tool.")
+mark_as_advanced(OPENCV_PYTHON_PIP_METADATA_INSTALLER)
+if(OPENCV_PYTHON_PIP_METADATA_INSTALL)
+  if(WIN32)
+    set(NEW_LINE "\n\r")
+  else()
+    set(NEW_LINE "\n")
+  endif()
+  file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/METADATA "")
+  file(APPEND ${CMAKE_CURRENT_BINARY_DIR}/METADATA "Metadata-Version: 2.1${NEW_LINE}")
+  file(APPEND ${CMAKE_CURRENT_BINARY_DIR}/METADATA "Name: opencv-python${NEW_LINE}")
```

Comment:
> > `opencv-python`
> 
> It is just a name of the package from dedicated repository: https://github.com/opencv/opencv-python
It is not an entity of this upstream repository. Perhaps we should not have it here.
> 
> Moreover there are several conflicting subpackages from "opencv-python", like headless or contrib variants.
> Ubuntu's python3-opencv package also doesn't know anything about opencv-python names:

Ack, I see why you may not want CMake's upstream to enter in this details that are mostly related to distributions (and mixing two different distribution, the base apt/conda system and pip).

> Modern Python uses PKG-INFO: https://www.python.org/dev/peps/pep-0314/

I am not an expert, I got the METADATA info from official Python documentation (and from PyPI's `opencv-python` installations):
* https://packaging.python.org/specifications/recording-installed-packages/
* https://packaging.python.org/en/latest/specifications/core-metadata/#core-metadata

From a quick search, it seems that it was introduced by the following PEPs that amend/complete pep-0314:
* https://www.python.org/dev/peps/pep-0376/
* https://www.python.org/dev/peps/pep-0627/

> > tries to install a Python library that depends on `opencv-python`
> 
> Are there another ways to mask package installation in pip?

That I am aware of, just passing to pip the `--no-deps`, that however do not install any dependency at and needs to be specified manually by users.

Reply:
@traversaro Thank you for information!

I have looked a bit on this specification: https://www.python.org/dev/peps/pep-0345/

There is a field to declare package name aliases (which wold work for us): [`Provides-Dist`](https://packaging.python.org/en/latest/specifications/core-metadata/#provides-dist-multiple-use)
E.g. we could have package name `opencv` and define known package aliases with names: `opencv-python`, `opencv-python-headless`.
But, looks like installation tools don't care about this field: https://stackoverflow.com/questions/61603438/do-any-tools-support-provides-dist-metadata-for-python-packages-if-not-why

Reply:
I am a random dude, who would like this feature to be implemented.

Regarding the package name, how about just making the `OPENCV_PYTHON_PIP_METADATA_INSTALL` option accept a String instead of a boolean and use that name everywhere `opencv_python`/`opencv-python` currently appears (with appropriate `_`/`-` translations)? This way, downstream package distributors could choose the appropriate name themselves?

Reply:
Just a curiosity, what is the use case for you for using this feature?

Reply:
I use ArchLinux and I normally prefer to install my python packages through pacman/AUR. The official `python-opencv` arch package builds and installs `opencv` from source (including the python bindings). This leads to the problems described by @traversaro, where `pip` doesn't know, that `opencv` is already installed. Currently, I roll my own PKGBUILD that installs the appropriate `dist-info` alongside the system-wide `opencv` package, but this is obviously inconvenient.